camel-idea-plugin/src/main/java/org/apache/camel/idea/completion/extension/CamelJavaBeanReferenceSmartCompletion.java

```diff
@@ -0,0 +1,132 @@
+
+
+/**
+ * Camel Java bean reference smart completion for lookup bean methods.
+ * <pre>
+ *     Caret inside {@code bean(MyClass.class,"<caret>")} will look up all public methods for the
+ *     class and it's super classes. If the a bean method is marked with @Handle it will bold and
+ *     for deprecated methods will be with strike through.
+ * </pre>
+ */
+public class CamelJavaBeanReferenceSmartCompletion extends CompletionProvider<CompletionParameters> {
+
+    public static final String BEAN_CLASS_KEY = "beanClassKey";
+
+    @Override
+    protected void addCompletions(@NotNull CompletionParameters completionParameters, ProcessingContext processingContext, @NotNull CompletionResultSet completionResultSet) {
+        final PsiElement element = completionParameters.getPosition();
+        final PsiClass psiClass = (PsiClass) processingContext.get(BEAN_CLASS_KEY);
+
+        List<LookupElement> answer = new ArrayList<>();
+
+        Collection<PsiMethod> methods = getMethods(psiClass);
+
+        for (PsiMethod method : methods) {
+            boolean isPrivate = getCamelIdeaUtils().isOneOfModifierType(method, JvmModifier.PRIVATE, JvmModifier.ABSTRACT);
+
+            if (!method.isConstructor() && !isPrivate) {
+                String presentableMethod = getPresentableMethodWithParameters(method);
+                answer.add(buildLookupElement(method, presentableMethod));
+            }
+        }
+        // are there any results then add them
+        if (!answer.isEmpty()) {
+            String hackVal = element.getText();
+            hackVal = hackVal.substring(1, hackVal.indexOf(CompletionUtil.DUMMY_IDENTIFIER_TRIMMED));
+            completionResultSet.withPrefixMatcher(hackVal).addAllElements(answer);
+            completionResultSet.stopHere();
+        }
+    }
+
+    @NotNull
+    private LookupElement buildLookupElement(PsiMethod method, String presentableMethod) {
+        LookupElementBuilder builder = LookupElementBuilder.create(method);
+        builder = builder.withPresentableText(presentableMethod);
+        builder = builder.withTypeText(method.getContainingClass().getName(), true);
+        builder = builder.withIcon(AllIcons.Nodes.Method);
+        if (getCamelIdeaUtils().isAnnotatedWithHandler(method)) {
+            //@Handle methods are marked with
+            builder = builder.withBoldness(true);
+        }
+        if (method.isDeprecated()) {
+            // mark as deprecated
+            builder = builder.withStrikeoutness(true);
+        }
+        return  builder.withAutoCompletionPolicy(AutoCompletionPolicy.GIVE_CHANCE_TO_OVERWRITE);
+    }
+
+    private String getPresentableMethodWithParameters(PsiMethod method) {
+        final String parameters = getMethodParameters(method);
+        String presentableMethod = method.getName();
+        if (!parameters.isEmpty()) {
+            presentableMethod = String.format("%s(%s)", method.getName(), parameters);
+        }
+        return presentableMethod;
```

Comment:
Do you want to have an empty set of parantheses when no params are available?

Reply:
if no parameter, the string "parameters" is empty :-)